rasa/utils/tensorflow/models.py

```diff
@@ -47,6 +48,25 @@
 logger = logging.getLogger(__name__)
 
 
+def _merge_batch_outputs(
+    all_outputs: Dict[Text, Union[np.ndarray, Dict[Text, np.ndarray]]],
+    batch_output: Dict[Text, Union[np.ndarray, Dict[Text, np.ndarray]]],
+) -> Dict[Text, Union[np.ndarray, Dict[Text, np.ndarray]]]:
+    if not all_outputs:
+        return batch_output
+    for key, val in batch_output.items():
+        if isinstance(val, np.ndarray):
+            all_outputs[key] = np.concatenate(
+                [all_outputs[key], batch_output[key]], axis=0
+            )
+
+        elif isinstance(val, dict):
+            # recurse and merge the inner dict first
+            all_outputs[key] = _merge_batch_outputs(all_outputs[key], val)
```

Comment:
> what about turning this into a recursive inner helper function and leaving the rest as non-recursive?

Not sure what you mean here 🙈 

> from the "outside world", I don't have to know that it's recursive

Why do you have to know about it now as well?

>its type annotation also simplifies to Dict[Text, Dict[Text, np.ndarray]], right?

I think this is linked to your first observation which I didn't understand. Could you elaborate?

Reply:
So, what I mean is: As someone merely calling the function, I don't care whether it's internally implemented as recursive or not. But I think the current type annotation is complicated as a result of reflecting the particulars of the implementation. I have in mind something like this, which hides the recursiveness:
```python
def _merge_batch_outputs(
    all_outputs: Dict[Text, Dict[Text, np.ndarray]],
    batch_output: Dict[Text, Dict[Text, np.ndarray]],
) -> Dict[Text, Dict[Text, np.ndarray]]:
    if not all_outputs:
        return batch_output

    def _recurse(
        _all_outputs: Dict[Text, Union[np.ndarray, Dict[Text, np.ndarray]]], 
        _batch_output: Dict[Text, Union[np.ndarray, Dict[Text, np.ndarray]]]
    ) -> Dict[Text, Union[np.ndarray, Dict[Text, np.ndarray]]]:
        for key, val in _batch_output.items():
            if isinstance(val, np.ndarray):
                _all_outputs[key] = np.concatenate(
                    [_all_outputs[key], _batch_output[key]], axis=0
                )

            elif isinstance(val, dict):
                # recurse and merge the inner dict first
                _all_outputs[key] = _recurse(all_outputs[key], val)

        return _all_outputs

    return _recurse(all_outputs, batch_output)
```

Reply:
But the type annotation of `all_outputs` in your method will be wrong then. The passed input will be `Dict[Text, Union[np.ndarray, Dict[Text, np.ndarray]]`

Reply:
Also, the recursion of `_recurse` method will not terminate I think.

Reply:
Ah, that's what I wasn't sure about. I thought the `Dict[Text, np.ndarray]` case was added only because of the recursive calls.

Reply:
The complete annotation comes because we expect the batch outputs to contain nested dicts for 'some' keys only. This structure will get reflected in the type annotation of the arguments anyhow.

> As someone merely calling the function, I don't care whether it's internally implemented as recursive or not

I still don't get how the caller needs to know about the recursion :) Making a call like `rasa_model._merge_batch_outputs(all_outputs, batch_output) does not reveal anything about the implementation, no?

Reply:
Yeah, it was all about me assuming that the annotation could be simplified if we didn't call the function from within itself. My bad. Basically, I thought that the way the function was implemented was affecting its annotation (which would've been undesirable).